### PR TITLE
[docs] Detailed type definitions

### PR DESCRIPTION
**Pull request by vladmoroz**
**Link**
https://deploy-preview-818--base-ui.netlify.app/new/components/dialog

**What's up**
- Fully-fledged, customisable type definitions in the API reference
- Data attribute tables
- Improved inline code highlighting
- Much cleaner prop definitions JSON

**How it works**
- Temporary prop definitions and translations are created using `@mui-internal/api-docs-builder`
- Type data and descriptions are pieced together into a single file per component. Unnecessary fields aren't included.
- Type overrides are looked up in `docs/reference/overrides`
  - Overrides common to many components are stored in `docs/reference/overrides/common.json`
  - Per-component overrides are stored in the corresponding files in that folder.
- Temporary data is deleted, new JSON is written into `docs/reference/generated`

`docs/data` folder is still in place and is used by the existing setup, but will be removed once we are finished.

Things that we will need to add to overrides pretty much all the time:
- Custom function types, because the original types are no good
- Data attributes documentation

**Screenshot**

image

### CODE REVIEW

File: docs/reference/generated/accordion-header.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "AccordionHeader",
+  "description": "",
```
Comment (vladmoroz) [inline docs/reference/generated/accordion-header.json:3]: Empty descriptions throw warnings during the API generation

File: packages/mui-base/src/Dialog/Backdrop/DialogBackdrop.tsx [hunk 2]
```diff
@@ -24,6 +24,7 @@ const customStyleHookMapping: CustomStyleHookMapping<DialogBackdrop.OwnerState>
 };
 
 /**
+ * An overlay displayed beneath the popup. Renders a `<div>` element.
```
Comment (vladmoroz) [inline packages/mui-base/src/Dialog/Backdrop/DialogBackdrop.tsx:27]: Do we like this and the other descriptions I added in the PR?

I plan to add/rework the descriptions for all the components in this style. (Most are missing).

Comment (vladmoroz) [inline packages/mui-base/src/Dialog/Backdrop/DialogBackdrop.tsx:27]: Do we like this and the other descriptions I added in the PR @colmtuite @samuelsycamore?

I plan to add/rework the descriptions for all the components in this style. (Most are missing).

File: docs/reference/overrides/common.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    "func": "function",
+    "node": "React.ReactNode",
+    "ref": "React.Ref",
+    "Array<number>": "number[]",
+    "Array<string>": "string[]",
+    "'bottom' | 'left' | 'right' | 'top'": "'top' | 'bottom' | 'left' | 'right'",
+    "'center' | 'end' | 'start'": "'start' | 'center' | 'end'",
+    "'both' | 'none' | 'x' | 'y'": "'none' | 'x' | 'y' | 'both'"
+  }
+}
```
Comment (vladmoroz) [inline docs/reference/overrides/common.json:42]: Point of interest

File: docs/reference/overrides/dialog-popup.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+      "description": "Present when the dialog is closed."
+    },
+    "[data-entering]": {
+      "description": "Present when the dialog is animating in."
+    },
+    "[data-exiting]": {
+      "description": "Present when the dialog is animating out."
+    }
+  }
+}
```
Comment (vladmoroz) [inline docs/reference/overrides/dialog-popup.json:17]: Point of interest: this is how data attributes are documented

File: docs/reference/overrides/number-field-root.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  "name": "NumberFieldRoot",
+  "props": {
+    "format": {
+      "type": "Intl.NumberFormatOptions"
+    },
+    "onValueChange": {
+      "type": "(value, event) => void"
+    }
+  }
+}
```
Comment (vladmoroz) [inline docs/reference/overrides/number-field-root.json:11]: Point of interest: this is how generated type definitions are overridden when they are no good

**Review by atomiks — APPROVED** [review]

File: packages/mui-base/src/Dialog/Backdrop/DialogBackdrop.tsx (hunk 2 above)
Comment (samuelsycamore) [reply packages/mui-base/src/Dialog/Backdrop/DialogBackdrop.tsx:27]: I left a note on one but in general I'm happy with the descriptions and their style. 🤝

**Review by colmtuite — APPROVED** [review]

**Pull request merged**